File: 백준/Silver/1965. 상자넣기/상자넣기.py
N = int(input())
boxes = list(map(int, input().split()))

# 가장 긴 증가하는 부분 수열 문제랑 같음
dp = [1] * N
# dp[i] : i를 마지막으로 갖는 부분 수열 중 길이의 최대값

for i in range(N):
    for j in range(i):
        if boxes[i] > boxes[j]:
            dp[i] = max(dp[i], dp[j]+1)

print(max(dp))

File: 백준/Silver/1965. 상자넣기/상자넣기.java
import java.util.*;
import java.io.*;
public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        int[] boxes = new int[N];
        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            boxes[i] = Integer.parseInt(st.nextToken());
        }
        int[] dp = new int[N];
        Arrays.fill(dp, 1);
        int answer = 0;

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < i; j++) {
                if (boxes[i] > boxes[j]) {
                    dp[i] = Math.max(dp[i], dp[j]+1);
                }
            }
            answer = Math.max(dp[i], answer);
        }

        System.out.print(answer);

    }
}
